packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/internal/KtorNetworkTransport.kt

```diff
@@ -0,0 +1,199 @@
+                                }
+                            }
+                    }
+
+                    addBody(method, body)
+                    addMethod(method)
+                }
+
+                try {
+                    when (method) {
+                        "delete" -> client.delete<HttpResponse>(url, requestBuilderBlock)
+                        "patch" -> client.patch<HttpResponse>(url, requestBuilderBlock)
+                        "post" -> client.post<HttpResponse>(url, requestBuilderBlock)
+                        "put" -> client.put<HttpResponse>(url, requestBuilderBlock)
+                        "get" -> client.get<HttpResponse>(url, requestBuilderBlock)
+                        else -> throw IllegalArgumentException("Wrong request method: '$method'")
+                    }.let {
+                        processHttpResponse(it)
+                    }
+                } catch (e: ServerResponseException) {
+                    // 500s are thrown as ServerResponseException
+                    processHttpResponse(e.response)
+                } catch (e: CancellationException) {
+                    println("EXCEPTION 1: $e")
+                    println("MESSAGE 1: ${e.message}")
+                    Response(-1, -1, mapOf(), "")
+                } catch (e: Exception) {
+                    println("EXCEPTION 2: $e")
+                    println("MESSAGE 2: ${e.message}")
+                    Response(-2, -2, mapOf(), "")
+                }
+            }
+        } catch (e: Exception) {
+            println("EXCEPTION 3: $e")
+            println("MESSAGE 3: ${e.message}")
+            return Response(-3, -3, mapOf(), "")
+        }
+    }
+
+    private suspend fun processHttpResponse(response: HttpResponse): Response {
+        val responseBody = response.receive<String>()
+        val responseStatusCode = response.status.value
+        val responseHeaders = parseHeaders(response.headers)
+        return httpResponse(responseStatusCode, responseHeaders, responseBody)
+    }
+
+    private fun HttpRequestBuilder.addBody(method: String, body: String) {
+        when (method) {
+            "delete", "patch", "post", "put" -> this.body = body
+        }
+    }
+
+    private fun HttpRequestBuilder.addMethod(method: String) {
+        when (method) {
+            "delete" -> this.method = HttpMethod.Delete
+            "patch" -> this.method = HttpMethod.Patch
+            "post" -> this.method = HttpMethod.Post
+            "put" -> this.method = HttpMethod.Put
+            "get" -> this.method = HttpMethod.Get
+        }
+    }
+
+    private fun getClient(): HttpClient {
+        // Need to freeze value as it is used inside the client's init lambda block, which also
+        // freezes captured objects too, see:
+        // https://youtrack.jetbrains.com/issue/KTOR-1223#focus=Comments-27-4618681.0-0
+        val frozenTimeout = timeoutMs.freeze()
+        return HttpClient(CIO) {
```

Comment:
Where is CIO defined?

Reply:
Apparently in `package io.ktor.client.engine.ci` 😅

Reply:
Removed the actual client, as ktor will resolve it automatically from the available clients at runtime. Added a TODO about fixating it, so we can control it even though people provide additional dependencies.